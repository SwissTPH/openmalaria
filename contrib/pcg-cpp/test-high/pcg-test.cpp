/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
