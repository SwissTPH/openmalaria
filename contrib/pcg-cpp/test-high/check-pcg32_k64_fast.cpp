#define RNG pcg32_k64_fast
#define TWO_ARG_INIT 0

#include "pcg-test.cpp"

