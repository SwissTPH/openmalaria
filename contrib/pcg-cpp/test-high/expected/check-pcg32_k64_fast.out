pcg32_k64_fast:
      -  result:      32-bit unsigned int
      -  period:      2^2112
      -  size:        264 bytes

Round 1:
  32bit: 0x623d127c 0xfa4987c9 0x1a27eeee 0x603880d8 0xd94c8039 0x0f4aa323
  Again: 0x623d127c 0xfa4987c9 0x1a27eeee 0x603880d8 0xd94c8039 0x0f4aa323
  Coins: HTHTTTHHHHTHHHTTHHTHHTTTHTHHTHTHTTHHTTTTHTTHTHHHTTTHTTTTTTTHHHHHT
  Rolls: 3 2 2 5 4 5 5 2 6 2 2 1 3 5 4 5 4 4 2 6 2 3 3 5 5 1 3 3 6 1 5 2 5
   -->   rolling dice used 33 random numbers
  Cards: 5d 3s 9c Jd 6c 8d 2s 8h Qd 9d Kd 3h 7d 4c 7s Jc Js Th 5s 2d Ad 9h
	 Ts Ks Ah 8c 3d 9s Qh 3c 7c 5c 6h 2h As Kh Td 7h 4h Jh Qs Kc 2c 6d
	 8s Ac Tc 4s 5h 4d Qc 6s

Round 2:
  32bit: 0x8df2a75a 0x8d0158a9 0xafeb1142 0x79c26727 0x149f2c2c 0xfdb241eb
  Again: 0x8df2a75a 0x8d0158a9 0xafeb1142 0x79c26727 0x149f2c2c 0xfdb241eb
  Coins: HHHTTHTHTTTTHTHHHTTHTTHHTTTHHTTTTHHHHHTHTTHTTTTHHTHHTHHTHHTHHHHHT
  Rolls: 6 4 1 4 4 6 6 4 2 3 2 4 4 1 6 6 6 2 4 3 2 1 4 4 2 2 3 4 4 2 4 6 2
   -->   rolling dice used 33 random numbers
  Cards: 6h Ts Qd 8c Ks 9d 2d Kc 4h Qc Td 5s 3d 8s 2s 6c Ah Kd Ac 5c Ad 3h
	 2c 5d Jc Qh 8d 3s As 2h 7h 7c 9s 5h 4d Kh 9c 6d 7s Th Qs Js 3c 9h
	 8h Jd Jh 7d Tc 4c 6s 4s

Round 3:
  32bit: 0x4d5299d6 0x7993beb4 0xa9878cdf 0xb23d608f 0x442110c0 0x701b5b8c
  Again: 0x4d5299d6 0x7993beb4 0xa9878cdf 0xb23d608f 0x442110c0 0x701b5b8c
  Coins: TTTHTTTHTHHTTHHTHTTTHHTTTHTTHTTHTTHTHTTHHHTTTHHTHHTHHHTHTHTTHTHTH
  Rolls: 3 3 1 3 2 4 2 1 6 3 4 4 5 6 1 4 2 5 1 2 3 4 2 6 2 5 3 4 4 6 6 1 6
   -->   rolling dice used 33 random numbers
  Cards: Th Kc 3h Qc 3d 7s Ks 2h Qd 4d 6c Ts 3s Kh 9h Ac Ad 5d 2d 4s 7h Jc
	 5h 9d 9c 8h Js 7d Jd Kd 7c 3c Jh 5c 8d 9s 8c 8s Tc 6h 5s Td Qh Ah
	 4h 4c As 6d 2c Qs 2s 6s

Round 4:
  32bit: 0xb3686885 0xbfa94438 0x47b29921 0x2787b39e 0x685c3666 0x844a798f
  Again: 0xb3686885 0xbfa94438 0x47b29921 0x2787b39e 0x685c3666 0x844a798f
  Coins: THHHTTHTHHHHHTTHTHTTTTHHTHTHTHTTHTHHHHHTTHHHHTHTHHTTHHTHTTHHHTHTH
  Rolls: 3 6 6 6 2 6 5 2 1 4 4 4 3 6 6 2 2 5 6 3 3 2 1 1 4 3 3 4 2 2 4 1 1
   -->   rolling dice used 33 random numbers
  Cards: 9h 2s Tc As Th 7d Jh 3h Td 7h 9s 4c 5s 5h 6h Jd Ts 2d Qd Ad Js Ks
	 8c 9c 3s 9d 8d 7s 4s Qc 6c 3c Jc 5d Ac 2h 7c Qs 8h Kd 8s 4h Kc 6s
	 Kh Ah 2c 4d 6d Qh 3d 5c

Round 5:
  32bit: 0xcf9ef109 0xcdc3a236 0xa3107d4a 0x8481cc76 0x14cb0dc0 0xbeb351d3
  Again: 0xcf9ef109 0xcdc3a236 0xa3107d4a 0x8481cc76 0x14cb0dc0 0xbeb351d3
  Coins: TTTTTTTTHTHHTTHTHTHTTTHHTHTHTTHHTHHTHHHHHTTHTTTHTTTHHTTTTTTTHTHHT
  Rolls: 2 1 3 5 3 2 6 3 5 5 1 3 2 5 5 2 5 2 4 5 4 5 1 3 1 6 1 6 5 2 1 2 6
   -->   rolling dice used 33 random numbers
  Cards: 8s 6s 3d 2c 9c 4h 5d 9s 7s 8c Ah 8d Td 5c Js 8h Qd 9h 2s Jc 6h As
	 Qh 4s Qc Kh Ts Qs Th 5s Ad Tc 9d 3s 4d 7h Kc 3c 7c 2d Ac 6c 4c Ks
	 3h 6d Kd Jh 2h 5h 7d Jd

